### ISSUE TITLE
Restore with parallel download of nuget packages

### ISSUE THREAD

**Issue by lawrencegripper** [fsprojects/Paket#2203]
### Description

I'm was just about to dig into the code to see if this was possible and wanted to check in first.

Would you be happy to accept a PR which looked at making Paket download nuget packages in parallel? Is there a reason this isn't done or is it already done and I just haven't found it?

I believe, as the paket.lock has already done all the resolution, I should be able to do this quite effectively but I'm yet to dig into it in detail. The motivation is that our build box has pretty good net connectivity and we'd love to saturate the pipe during the restore operation to speed things up. 

I can't guarantee I'll be able to pull it off as I'm new to F# but happy to have a go.

Ps. We recently moved a large solution over to paket and it has saved us countless hours of pain with nuget dependency clashes - thanks for all the hard work!

**Comment by forki** [comment]
IIRC we already restore in parallel. But maybe we broke it

Am 27.03.2017 20:03 schrieb "Lawrence Gripper" <notifications@github.com>:

**Comment by lawrencegripper** [comment]
I had a chance to dig into this a bit more and took a fiddler trace of what I see, it looks like the first 8 files are received in parallel. The remaining files download serially, below is the timeline from the fiddler trace. 

![image](https://cloud.githubusercontent.com/assets/1939288/25487268/9d42f1b4-2b5b-11e7-8635-22b30733a9a6.png)

Looking at the code (first time to F# so apologies if I'm miles off):

[CreateModel](https://github.com/fsprojects/Paket/blob/165f0cfd0be7c69b457c5e59f4699aec988320f7/src/Paket.Core/Installation/InstallProcess.fs#L184)
Calls: CreateInstallModel
Calls: ExtractPackage
Calls: DownloadPakage

My understanding of Async.Parallel in CreateModel is that a task should be created for each item in the lock group and the those tasks are grouped into an array and awaited with the Async.RunSynchronously so I'm pretty surprised with the pattern I'm seeing. 

Next step is to get out the debugger and learn a bit more F#, wanted to share my investigation so far in case there is anything obvious that I've missunderstood which would explain the behavior.  

Thanks again for all the work on Paket, really enjoying it!

**Comment by baronfel** [comment]
we need some sort of coordinator to manage a set number of asyncronous operations in flight at a time, which seems like an ideal use case for an agent like this: http://www.fssnip.net/nX/title/Limit-degree-of-parallelism-using-an-agent

**Comment by cloudRoutine** [comment]
I think the bigger issue is that it doesn't seem to be retrieving packages from the nuget cache instead of downloading new ones

**Comment by matthid** [comment]
I'm not sure if this maybe looks like this because some cap (for example on parallel downloads). First all start and then for each new slow another one starts?
Anyway please test against latest.
